pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -430,6 +438,164 @@ func consumerRefMatches(consumer *corev1.ObjectReference, machine *machinev1.Mac
+		// userDataSecret is the pointer ignition originally
+		// created by openshift installer. Fetch it to extract
+		// its content (an URL containg full ignition and a
+		// CA cert for the URL's https connection)
+		userDataSecret := corev1.Secret{}
+		key = client.ObjectKey{
+			Name:      config.UserData.Name,
+			Namespace: config.UserData.Namespace,
+		}
+		err = a.client.Get(ctx, key, &userDataSecret)
+		if err != nil {
+			log.Printf("Cannot get %s secret: %w", config.UserData.Name, err)
+			return err
+		}
+		encodedUserDataBytes := userDataSecret.Data["userData"]
+		var userData []byte
+		userData, err = base64.StdEncoding.DecodeString(string(encodedUserDataBytes))
+		if err != nil {
+			log.Printf("Error base64 decoding userData : %s", err.Error())
+			return fmt.Errorf("error base64 decoding userData: %w", err)
+		}
+
+		// Define an ignition data struct to unmarshal incoming json
+		type IgnitionData struct {
+			Ignition struct {
+				Config struct {
+					Append []struct {
+						Source string
+					}
+				}
+				Security struct {
+					TLS struct {
+						CertificateAuthorities []struct {
+							Source string
+						}
+					}
+				}
+			}
+		}
+		var ignData IgnitionData
+
+		if err := json.Unmarshal(userData, &ignData); err != nil {
+			log.Printf("Cannot unmarshal ignition config from userData : %s, %w", config.UserData.Name, err)
+			return err
+		}
+
+		var ignitionURL, caCertRaw string
+		// Fetch IgnitionURL and the CA cert
+		if len(ignData.Ignition.Config.Append) > 0 && len(ignData.Ignition.Config.Append[0].Source) > 0 {
+			ignitionURL = ignData.Ignition.Config.Append[0].Source
+		} else {
+			// We didn't get any URL
+			err := fmt.Errorf("error extracting full Ignition URL")
+			log.Printf("URL not found : %s", err.Error())
+			return err
+		}
+
+		if len(ignData.Ignition.Security.TLS.CertificateAuthorities) > 0 && len(ignData.Ignition.Security.TLS.CertificateAuthorities[0].Source) > 0 {
+			caCertRaw = ignData.Ignition.Security.TLS.CertificateAuthorities[0].Source
+		} else {
+			// We didn't get the URL's Cert
+			err := fmt.Errorf("CA cert for full Ignition URL was not specified")
+			log.Printf("Cert not found in ignition data: %s", err.Error())
+			return err
+		}
+		caCertB64 := strings.TrimPrefix(caCertRaw, "data:text/plain;charset=utf-8;base64,")
+
+		caCert, err := base64.StdEncoding.DecodeString(caCertB64)
+		if err != nil {
+			log.Printf("Could not base64 decode caCert: %s", err.Error())
+			return fmt.Errorf("could not base64 decode caCert: %w", err)
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		client := retryablehttp.NewClient()
+		client.HTTPClient.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: caCertPool,
+			},
+		}
+
+		// Get the full ignition from the URL
+		resp, err := client.Get(ignitionURL)
+		if err != nil || resp.StatusCode != http.StatusOK {
+			log.Printf("Error fetching full ignition from URL : %s", err.Error())
```

Comment:
If the response is non-200, `err` may still be `nil`. I think it's worth a dedicated log statement for that situation, and you may need to modify the error being returned below.

Reply:
+1. Made the change. Will add it to the other requested changes.